### PR TITLE
Merge basic and full markdown renderer

### PR DESCRIPTION
**Pull request by mrdrogdrog**
### Component/Part
Markdown it renderer

### Description
The original idea of the basic-markdown-renderer and the full-markdown-renderer was to reduce the complexity. The basic markdown renderer should just render markdown code and the full markdown renderer should implement all the special hedgedoc stuff like the embeddings.
While developing other aspects of the software I noticed, that it makes more sense to split the markdown-renderer by the view and not by the features. E.g.: The slide markdown renderer must translate `<hr>` into `<sections>` for the slides and the document markdown renderer must provide precise scroll positions. But both need e.g. the ability to show a youtube video.

Therefore this PR merges the two renderers into one (and improve some other pieces of code), so we can split them into the document-markdown-renderer and the slide-markdown-renderer.

### Steps

- [x] Added implementation
- [x] Added / updated tests
- [ ] Added / updated documentation
- [ ] Extended changelog
- [x] I read the [contribution documentation](https://github.com/hedgedoc/react-client/blob/main/CONTRIBUTING.md) and signed-off my commits to accept the DCO.

### CODE REVIEW

File: src/components/markdown-renderer/basic-markdown-renderer.tsx [hunk 1]
```diff
@@ -4,43 +4,105 @@
...
+    }
+  })
 
   return (
-    <div className={ `${ className ?? '' } d-flex flex-column align-items-center` }>
-      <DocumentLengthLimitReachedAlert contentLength={ content.length }/>
-      <div ref={ documentReference } className={ 'markdown-body w-100 d-flex flex-column align-items-center' }>
+    <div ref={ outerContainerRef } className={ 'position-relative' }>
+      <InvalidYamlAlert show={ showYamlError }/>
+      <DocumentLengthLimitReachedAlert show={ isTrimmed }/>
```
Comment (ErikMichelson) [inline src/components/markdown-renderer/basic-markdown-renderer.tsx:103]: The variable name `isTrimmed` is a bit misleading here, as one could see isTrimmed as the note is already short. Maybe rename it to `exceededLimit`?

File: src/components/markdown-renderer/hooks/use-trimmed-content.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { useSelector } from 'react-redux'
+import { ApplicationState } from '../../../redux'
+import { useMemo } from 'react'
+
+export const useTrimmedContent = (content: string): [trimmedContent: string, isTrimmed: boolean] => {
+  const maxLength = useSelector((state: ApplicationState) => state.config.maxDocumentLength)
+  const isTrimmed = content.length > maxLength
```
Comment (ErikMichelson) [inline src/components/markdown-renderer/hooks/use-trimmed-content.ts:13]: As stated above, I'd rename this variable.

File: src/components/markdown-renderer/invalid-yaml-alert.tsx [hunk 3]
```diff
@@ -11,14 +11,14 @@ import { InternalLink } from '../common/links/internal-link'
 import { ShowIf } from '../common/show-if/show-if'
 
 export interface InvalidYamlAlertProps {
-  showYamlError: boolean
+  show: boolean
 }
```
Comment (ErikMichelson) [inline src/components/markdown-renderer/invalid-yaml-alert.tsx:15]: Can't we just get rid of all these interfaces that only contain a `show` boolean and instead use a common show-modal interface?

File: src/components/markdown-renderer/markdown-it-configurator/BasicMarkdownItConfigurator.tsx [hunk 4]
```diff
@@ -19,11 +19,77 @@ import { MarkdownItParserDebugger } from '../markdown-it-plugins/parser-debugger
...
+
+  public pushConfig(plugin: MarkdownIt.PluginSimple): this {
+    this.configurations.push(plugin)
+    return this
+  }
+
+  public buildConfiguredMarkdownIt(): MarkdownIt {
+    const markdownIt = new MarkdownIt('default', {
+      html: true,
+      breaks: true,
```
Comment (ErikMichelson) [inline src/components/markdown-renderer/markdown-it-configurator/BasicMarkdownItConfigurator.tsx:66]: As I just see this: Maybe add a TODO here that this (`breaks`) needs to be updated according to the frontmatter later?

File: yarn.lock [hunk 5]
```diff
@@ -15498,3 +15498,6 @@ zip-stream@^2.1.2:
     archiver-utils "^2.1.0"
     compress-commons "^2.1.1"
     readable-stream "^3.4.0"
+
+
+
```
Comment (ErikMichelson) [inline yarn.lock:15503]: Why?

File: src/components/editor-page/document-bar/share/share-modal.tsx [hunk 6]
```diff
@@ -39,7 +40,7 @@ export const ShareModal: React.FC<ShareModalProps> = ({ show, onHide }) => {
         <Trans i18nKey={ 'editor.modal.shareLink.editorDescription' }/>
         <CopyableField content={ `${ baseUrl }/n/${ id }?${ editorMode }` } nativeShareButton={ true }
                        url={ `${ baseUrl }/n/${ id }?${ editorMode }` }/>
-        <ShowIf condition={ noteFrontmatter.type === 'slide' }>
+        <ShowIf condition={ noteFrontmatter && noteFrontmatter.type === NoteType.SLIDE }>
```
[the remark this answers is not in the document]
Comment (ErikMichelson) [reply src/components/editor-page/document-bar/share/share-modal.tsx:43]: And if you want to change this, be sure to do it in line 48 too.

**Review by ErikMichelson — CHANGES_REQUESTED** [review]

File: yarn.lock (hunk 5 above)
Comment (mrdrogdrog) [reply yarn.lock:15503]: As you can see in the last commit I hoped to fix the ci with this work around.

File: src/components/editor-page/document-bar/share/share-modal.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (mrdrogdrog) [reply src/components/editor-page/document-bar/share/share-modal.tsx:43]: I removed the null check. Frontmatter can't be undefined.

File: src/components/editor-page/note-frontmatter/note-frontmatter.test.ts [hunk 7]
```diff
@@ -9,127 +9,102 @@ import MarkdownIt from 'markdown-it'
...
-    `,
-      {
-        breaks: false
-      },
-      {
-        breaks: false
-      })
+    `)
+
+
```
[the remark this answers is not in the document]
Comment (mrdrogdrog) [reply src/components/editor-page/note-frontmatter/note-frontmatter.test.ts:105]: removed

File: src/components/markdown-renderer/basic-markdown-renderer.tsx (hunk 1 above)
Comment (mrdrogdrog) [reply src/components/markdown-renderer/basic-markdown-renderer.tsx:103]: Renamed

File: src/components/markdown-renderer/hooks/use-trimmed-content.ts (hunk 2 above)
Comment (mrdrogdrog) [reply src/components/markdown-renderer/hooks/use-trimmed-content.ts:13]: same

File: src/components/markdown-renderer/invalid-yaml-alert.tsx (hunk 3 above)
Comment (mrdrogdrog) [reply src/components/markdown-renderer/invalid-yaml-alert.tsx:15]: Merged

File: src/components/markdown-renderer/markdown-it-configurator/BasicMarkdownItConfigurator.tsx (hunk 4 above)
Comment (mrdrogdrog) [reply src/components/markdown-renderer/markdown-it-configurator/BasicMarkdownItConfigurator.tsx:66]: I linked it to the redux state

File: yarn.lock (hunk 5 above)
Comment (mrdrogdrog) [reply yarn.lock:15503]: Works again. The problem was an import cycle.

File: src/components/editor-page/app-bar/help-button/cheatsheet-line.tsx [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+import React, { Suspense, useCallback } from 'react'
+import { WaitSpinner } from '../../../common/wait-spinner/wait-spinner'
+
+export interface CheatsheetLineProps {
+  code: string,
+  onTaskCheckedChange: (newValue: boolean) => void
+}
+
+
```
Comment (ErikMichelson) [inline src/components/editor-page/app-bar/help-button/cheatsheet-line.tsx:15]: Multiple blanks

File: src/components/markdown-renderer/hooks/use-trimmed-content.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+ * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { useSelector } from 'react-redux'
+import { ApplicationState } from '../../../redux'
+import { useMemo } from 'react'
+
+export const useTrimmedContent = (content: string): [trimmedContent: string, isTrimmed: boolean] => {
```
Comment (ErikMichelson) [inline src/components/markdown-renderer/hooks/use-trimmed-content.ts:11]: It's non-functional, but maybe change the variable name in the method type signature as well?

File: src/components/markdown-renderer/replace-components/markmap/markmap-frame.tsx [hunk 10]
```diff
@@ -57,7 +57,7 @@ export const MarkmapFrame: React.FC<MarkmapFrameProps> = ({ code }) => {
         console.error(error)
       }
     })
-                              .catch(() => {
+                                                               .catch(() => {
```
Comment (ErikMichelson) [inline src/components/markdown-renderer/replace-components/markmap/markmap-frame.tsx:60]: Technically that's correct due to our editor config.
What about moving the `.then()` after the `import` into the next line to have the `.catch` not intended that much?

File: src/components/markdown-renderer/replace-components/mermaid/mermaid-chart.tsx [hunk 11]
```diff
@@ -66,7 +67,7 @@ export const MermaidChart: React.FC<MermaidChartProps> = ({ code }) => {
         showError(message || t('renderer.mermaid.unknownError'))
       }
     })
-                     .catch(() => showError('Error while loading mermaid'))
+                                                      .catch(() => showError('Error while loading mermaid'))
```
Comment (ErikMichelson) [inline src/components/markdown-renderer/replace-components/mermaid/mermaid-chart.tsx:70]: Same as my comment above regarding the then-catch intendation.

File: src/components/editor-page/app-bar/help-button/help-modal.tsx [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        </Button>
+        <Button variant={ 'light' }
+                className={ `nav-link nav-item ${ tab === HelpTabStatus.Links ? 'active' : '' }` }
+                onClick={ () => setTab(HelpTabStatus.Links) }>
+          <Trans i18nKey={ 'editor.help.links.title' }/>
+        </Button>
+      </nav>
+      { tabContent }
+    </Modal.Body>
+  </Modal>
```
Comment (ErikMichelson) [inline src/components/editor-page/app-bar/help-button/help-modal.tsx:63]: As you already use common-modal props, couldn't you reduce unnecessary verbosity by using the `CommonModal` component?

Comment (mrdrogdrog) [reply src/components/editor-page/app-bar/help-button/help-modal.tsx:63]: Yes but this modal uses a custom header. Hm. Maybe if i extend the commonmodal with a subheadline.

File: src/components/editor-page/app-bar/help-button/cheatsheet-line.tsx (hunk 8 above)
Comment (mrdrogdrog) [reply src/components/editor-page/app-bar/help-button/cheatsheet-line.tsx]: fixed

File: src/components/markdown-renderer/hooks/use-trimmed-content.ts (hunk 9 above)
Comment (mrdrogdrog) [reply src/components/markdown-renderer/hooks/use-trimmed-content.ts]: forgot that. fixed

File: src/components/markdown-renderer/replace-components/markmap/markmap-frame.tsx (hunk 10 above)
Comment (mrdrogdrog) [reply src/components/markdown-renderer/replace-components/markmap/markmap-frame.tsx]: fixed

File: src/components/markdown-renderer/replace-components/mermaid/mermaid-chart.tsx (hunk 11 above)
Comment (mrdrogdrog) [reply src/components/markdown-renderer/replace-components/mermaid/mermaid-chart.tsx]: fixed

File: src/components/editor-page/app-bar/help-button/help-modal.tsx (hunk 12 above)
Comment (mrdrogdrog) [reply src/components/editor-page/app-bar/help-button/help-modal.tsx]: fixed

**Review by ErikMichelson — APPROVED** [review]

**Review by DerMolly — APPROVED** [review]

**Review by DerMolly — APPROVED** [review]

**Pull request merged**